pkg/agent/controller/agent.go

```diff
@@ -263,35 +384,86 @@ func (a *Controller) updateExportedServiceStatus(name, namespace string, condTyp
 	}
 }
 
-func serviceExportConditionEqual(c1, c2 *lighthousev2a1.ServiceExportCondition) bool {
-	return c1.Type == c2.Type && c1.Status == c2.Status && reflect.DeepEqual(c1.Reason, c2.Reason) &&
-		reflect.DeepEqual(c1.Message, c2.Message)
+func (a *Controller) serviceImportEquivalent(obj1, obj2 *unstructured.Unstructured) bool {
+	return syncer.DefaultResourcesEquivalent(obj1, obj2) &&
+		equality.Semantic.DeepEqual(util.GetNestedField(obj1, "status"), util.GetNestedField(obj2, "status"))
 }
 
-func (a *Controller) newClusterStatus(service *corev1.Service, clusterID string) lighthousev2a1.ClusterStatus {
-	mcsIp := getGlobalIpFromService(service)
-	if mcsIp == "" {
-		mcsIp = service.Spec.ClusterIP
-	}
+func (a *Controller) endpointEquivalent(obj1, obj2 *unstructured.Unstructured) bool {
+	return equality.Semantic.DeepEqual(util.GetNestedField(obj1, "subsets"),
+		util.GetNestedField(obj2, "subsets"))
+}
 
-	return lighthousev2a1.ClusterStatus{
-		Cluster: clusterID,
-		IPs:     []string{mcsIp},
-	}
+func serviceExportConditionEqual(c1, c2 *lighthousev2a1.ServiceExportCondition) bool {
+	return c1.Type == c2.Type && c1.Status == c2.Status && reflect.DeepEqual(c1.Reason, c2.Reason) &&
+		reflect.DeepEqual(c1.Message, c2.Message)
 }
 
 func (a *Controller) newServiceImport(svcExport *lighthousev2a1.ServiceExport) *lighthousev2a1.ServiceImport {
 	return &lighthousev2a1.ServiceImport{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: svcExport.Name + "-" + svcExport.Namespace + "-" + a.clusterID,
+			Name: a.getObjectNameWithClusterId(svcExport.Name, svcExport.Namespace),
 			Annotations: map[string]string{
 				originName:      svcExport.Name,
 				originNamespace: svcExport.Namespace,
 			},
+			Labels: map[string]string{
+				labelSourceName:      svcExport.Name,
+				labelSourceNamespace: svcExport.Namespace,
+				labelSourceCluster:   a.clusterID,
+			},
 		},
 	}
 }
 
+func (a *Controller) getServiceImport(svcExport *lighthousev2a1.ServiceExport) (*lighthousev2a1.ServiceImport, error) {
+	siName := a.getObjectNameWithClusterId(svcExport.Name, svcExport.Namespace)
+	serviceImport, err := a.lighthouseClient.LighthouseV2alpha1().ServiceImports(a.namespace).Get(siName, metav1.GetOptions{})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return serviceImport, nil
+}
+
+func (a *Controller) getIpsForService(service *corev1.Service, siType lighthousev2a1.ServiceImportType) []string {
+	if siType == lighthousev2a1.SuperclusterIP {
+		mcsIp := getGlobalIpFromService(service)
+		if mcsIp == "" {
+			mcsIp = service.Spec.ClusterIP
+		}
+
+		return []string{mcsIp}
+	}
+
+	endpoint, err := a.kubeClientSet.CoreV1().Endpoints(service.Namespace).Get(service.Name, metav1.GetOptions{})
+	if err != nil {
+		klog.V(log.DEBUG).Infof("Endpoints for svc  (%s/%s) not found: %v", service.Namespace, service.Name, err)
+		return make([]string, 0)
```

Comment:
If the error is `IsNotFound`, I think we should return something (nil should be fine)  to indicate to the caller to re-queue to be complete. I don't think we can assume that `endpointToRemoteServiceImport` will subsequently run.

Reply:
It is possible to bring up a service with 0 replicaset and scale it up later. `endpointToRemoteServiceImport` should run once endpoints are created. If they're not, we'd want to keep ipList to empty. In what scenarios will we get `NotFound` here and yet `endpointToRemoteServiceImport` won't run?

Reply:
Oh, I think you meant error is other than `NotFound`. Done that.

Reply:
> It is possible to bring up a service with 0 replicaset and scale it up later. `endpointToRemoteServiceImport` should run once endpoints are created. If they're not, we'd want to keep ipList to empty. In what scenarios will we get `NotFound` here and yet `endpointToRemoteServiceImport` won't run?

We shouldn't. I'm referring to an error other than `IsNotFound`, eg transient failure accessing the API server. The EP might already exist in which case `endpointToRemoteServiceImport` won't run and neither will this run again unless we re-queue (or unless some update happens to occur).  It's an edge case - we can handle it in a follow-up if you want so we can move this ahead.